Compilation issue for xdp-bench

I am compliling xdp-tools. First I run ./configure and the I run make. The compilation till xdp-dump happens correctly but after that I starts giving the error for xdp-bench. The following is the error for the same.

"""
In file included from xdp_redirect_basic.c:25:
xdp_redirect_basic.c: In function ‘sample_init’:
../lib/util/xdp_sample.h:72:21: error: implicit declaration of function ‘bpf_program__type’; did you mean ‘bpf_program__size’? [-Werror=implicit-function-declaration]
   72 |                 if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)\
      |                     ^~~~~~~~~~~~~~~~~
../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
  111 |                         __attach_tp(tp_xdp_redirect_map);          \
      |                         ^~~~~~~~~~~
xdp_redirect_basic.c:31:1: note: in expansion of macro ‘DEFINE_SAMPLE_INIT’
   31 | DEFINE_SAMPLE_INIT(xdp_redirect_basic);
      | ^~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
In file included from xdp_redirect_cpumap.c:28:
xdp_redirect_cpumap.c: In function ‘sample_init’:
../lib/util/xdp_sample.h:72:21: error: implicit declaration of function ‘bpf_program__type’; did you mean ‘bpf_program__size’? [-Werror=implicit-function-declaration]
   72 |                 if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)\
      |                     ^~~~~~~~~~~~~~~~~
../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
  111 |                         __attach_tp(tp_xdp_redirect_map);          \
      |                         ^~~~~~~~~~~
xdp_redirect_cpumap.c:55:1: note: in expansion of macro ‘DEFINE_SAMPLE_INIT’
   55 | DEFINE_SAMPLE_INIT(xdp_redirect_cpumap);
      | ^~~~~~~~~~~~~~~~~~
xdp_redirect_cpumap.c: In function ‘do_redirect_cpumap’:
xdp_redirect_cpumap.c:240:24: error: implicit declaration of function ‘bpf_object__next_program’; did you mean ‘bpf_object__unpin_programs’? [-Werror=implicit-function-declaration]
  240 |         while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~
      |                        bpf_object__unpin_programs
xdp_redirect_cpumap.c:240:22: error: assignment to ‘struct bpf_program *’ from ‘int’ makes pointer from integer without a cast [-Werror=int-conversion]
  240 |         while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
      |                      ^
xdp_redirect_cpumap.c:242:21: error: implicit declaration of function ‘bpf_program__expected_attach_type’; did you mean ‘bpf_program__set_expected_attach_type’? [-Werror=implicit-function-declaration]
  242 |                     bpf_program__expected_attach_type(prog) == BPF_XDP)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     bpf_program__set_expected_attach_type
cc1: all warnings being treated as errors
In file included from xdp_redirect_devmap.c:24:
xdp_redirect_devmap.c: In function ‘sample_init’:
../lib/util/xdp_sample.h:72:21: error: implicit declaration of function ‘bpf_program__type’; did you mean ‘bpf_program__size’? [-Werror=implicit-function-declaration]
   72 |                 if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)\
      |                     ^~~~~~~~~~~~~~~~~
../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
  111 |                         __attach_tp(tp_xdp_redirect_map);          \
      |                         ^~~~~~~~~~~
xdp_redirect_devmap.c:30:1: note: in expansion of macro ‘DEFINE_SAMPLE_INIT’
   30 | DEFINE_SAMPLE_INIT(xdp_redirect_devmap);
      | ^~~~~~~~~~~~~~~~~~
xdp_redirect_devmap.c: In function ‘do_redirect_devmap’:
xdp_redirect_devmap.c:72:24: error: implicit declaration of function ‘bpf_object__next_program’; did you mean ‘bpf_object__unpin_programs’? [-Werror=implicit-function-declaration]
   72 |         while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~
      |                        bpf_object__unpin_programs
xdp_redirect_devmap.c:72:22: error: assignment to ‘struct bpf_program *’ from ‘int’ makes pointer from integer without a cast
-Werror=int-conversion]
   72 |         while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
      |                      ^
xdp_redirect_devmap.c:74:21: error: implicit declaration of function ‘bpf_program__expected_attach_type’; did you mean ‘bpf_program__set_expected_attach_type’? [-Werror=implicit-function-declaration]
   74 |                     bpf_program__expected_attach_type(prog) == BPF_XDP)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     bpf_program__set_expected_attach_type
cc1: all warnings being treated as errors
In file included from xdp_redirect_devmap_multi.c:27:
xdp_redirect_devmap_multi.c: In function ‘sample_init’:
../lib/util/xdp_sample.h:72:21: error: implicit declaration of function ‘bpf_program__type’; did you mean ‘bpf_program__size’? [-Werror=implicit-function-declaration]
   72 |                 if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)\
      |                     ^~~~~~~~~~~~~~~~~
../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
  111 |                         __attach_tp(tp_xdp_redirect_map);          \
      |                         ^~~~~~~~~~~
xdp_redirect_devmap_multi.c:37:1: note: in expansion of macro ‘DEFINE_SAMPLE_INIT’
   37 | DEFINE_SAMPLE_INIT(xdp_redirect_devmap_multi);
      | ^~~~~~~~~~~~~~~~~~
xdp_redirect_devmap_multi.c: In function ‘do_redirect_devmap_multi’:
xdp_redirect_devmap_multi.c:108:24: error: implicit declaration of function ‘bpf_object__next_program’; did you mean ‘bpf_object__unpin_programs’? [-Werror=implicit-function-declaration]
  108 |         while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~
      |                        bpf_object__unpin_programs
xdp_redirect_devmap_multi.c:108:22: error: assignment to ‘struct bpf_program *’ from ‘int’ makes pointer from integer without a cast [-Werror=int-conversion]
  108 |         while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
      |                      ^
xdp_redirect_devmap_multi.c:110:21: error: implicit declaration of function ‘bpf_program__expected_attach_type’; did you mean
bpf_program__set_expected_attach_type’? [-Werror=implicit-function-declaration]
  110 |                     bpf_program__expected_attach_type(prog) == BPF_XDP)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     bpf_program__set_expected_attach_type
cc1: all warnings being treated as errors
In file included from xdp_basic.c:25:
xdp_basic.c: In function ‘sample_init’:
../lib/util/xdp_sample.h:72:21: error: implicit declaration of function ‘bpf_program__type’; did you mean ‘bpf_program__size’? [-Werror=implicit-function-declaration]
   72 |                 if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)\
      |                     ^~~~~~~~~~~~~~~~~
../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
  111 |                         __attach_tp(tp_xdp_redirect_map);          \
      |                         ^~~~~~~~~~~
xdp_basic.c:30:1: note: in expansion of macro ‘DEFINE_SAMPLE_INIT’
   30 | DEFINE_SAMPLE_INIT(xdp_basic);
      | ^~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [../lib/common.mk:107: xdp-bench] Error 1
make: *** [Makefile:34: xdp-bench] Error 2
"""

how to resolve this issue
![image](https://github.com/xdp-project/xdp-tools/assets/58917332/fba6ef9f-4890-4798-a2f9-ae7d7c131474)

What's the output of ./configure?

On Tuesday, May 30, 2023, Keshav ***@***.***> wrote:
> I am compliling xdp-tools. First I run ./configure and the I run make.
The compilation till xdp-dump happens correctly but after that I starts
giving the error for xdp-bench. The following is the error for the same.
>
> """
> In file included from xdp_redirect_basic.c:25:
> xdp_redirect_basic.c: In function ‘sample_init’:
> ../lib/util/xdp_sample.h:72:21: error: implicit declaration of function
‘bpf_program__type’; did you mean ‘bpf_program__size’?
[-Werror=implicit-function-declaration]
> 72 | if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)
> | ^~~~~~~~~~~~~~~~~
> ../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
> 111 | __attach_tp(tp_xdp_redirect_map);
> | ^~~~~~~~~~~
> xdp_redirect_basic.c:31:1: note: in expansion of macro
‘DEFINE_SAMPLE_INIT’
> 31 | DEFINE_SAMPLE_INIT(xdp_redirect_basic);
> | ^~~~~~~~~~~~~~~~~~
> cc1: all warnings being treated as errors
> In file included from xdp_redirect_cpumap.c:28:
> xdp_redirect_cpumap.c: In function ‘sample_init’:
> ../lib/util/xdp_sample.h:72:21: error: implicit declaration of function
‘bpf_program__type’; did you mean ‘bpf_program__size’?
[-Werror=implicit-function-declaration]
> 72 | if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)
> | ^~~~~~~~~~~~~~~~~
> ../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
> 111 | __attach_tp(tp_xdp_redirect_map);
> | ^~~~~~~~~~~
> xdp_redirect_cpumap.c:55:1: note: in expansion of macro
‘DEFINE_SAMPLE_INIT’
> 55 | DEFINE_SAMPLE_INIT(xdp_redirect_cpumap);
> | ^~~~~~~~~~~~~~~~~~
> xdp_redirect_cpumap.c: In function ‘do_redirect_cpumap’:
> xdp_redirect_cpumap.c:240:24: error: implicit declaration of function
‘bpf_object__next_program’; did you mean ‘bpf_object__unpin_programs’?
[-Werror=implicit-function-declaration]
> 240 | while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
> | ^~~~~~~~~~~~~~~~~~~~~~~~
> | bpf_object__unpin_programs
> xdp_redirect_cpumap.c:240:22: error: assignment to ‘struct bpf_program *’
from ‘int’ makes pointer from integer without a cast
[-Werror=int-conversion]
> 240 | while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
> | ^
> xdp_redirect_cpumap.c:242:21: error: implicit declaration of function
‘bpf_program__expected_attach_type’; did you mean
‘bpf_program__set_expected_attach_type’?
[-Werror=implicit-function-declaration]
> 242 | bpf_program__expected_attach_type(prog) == BPF_XDP)
> | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> | bpf_program__set_expected_attach_type
> cc1: all warnings being treated as errors
> In file included from xdp_redirect_devmap.c:24:
> xdp_redirect_devmap.c: In function ‘sample_init’:
> ../lib/util/xdp_sample.h:72:21: error: implicit declaration of function
‘bpf_program__type’; did you mean ‘bpf_program__size’?
[-Werror=implicit-function-declaration]
> 72 | if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)
> | ^~~~~~~~~~~~~~~~~
> ../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
> 111 | __attach_tp(tp_xdp_redirect_map);
> | ^~~~~~~~~~~
> xdp_redirect_devmap.c:30:1: note: in expansion of macro
‘DEFINE_SAMPLE_INIT’
> 30 | DEFINE_SAMPLE_INIT(xdp_redirect_devmap);
> | ^~~~~~~~~~~~~~~~~~
> xdp_redirect_devmap.c: In function ‘do_redirect_devmap’:
> xdp_redirect_devmap.c:72:24: error: implicit declaration of function
‘bpf_object__next_program’; did you mean ‘bpf_object__unpin_programs’?
[-Werror=implicit-function-declaration]
> 72 | while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
> | ^~~~~~~~~~~~~~~~~~~~~~~~
> | bpf_object__unpin_programs
> xdp_redirect_devmap.c:72:22: error: assignment to ‘struct bpf_program *’
from ‘int’ makes pointer from integer without a cast
> -Werror=int-conversion]
> 72 | while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
> | ^
> xdp_redirect_devmap.c:74:21: error: implicit declaration of function
‘bpf_program__expected_attach_type’; did you mean
‘bpf_program__set_expected_attach_type’?
[-Werror=implicit-function-declaration]
> 74 | bpf_program__expected_attach_type(prog) == BPF_XDP)
> | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> | bpf_program__set_expected_attach_type
> cc1: all warnings being treated as errors
> In file included from xdp_redirect_devmap_multi.c:27:
> xdp_redirect_devmap_multi.c: In function ‘sample_init’:
> ../lib/util/xdp_sample.h:72:21: error: implicit declaration of function
‘bpf_program__type’; did you mean ‘bpf_program__size’?
[-Werror=implicit-function-declaration]
> 72 | if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)
> | ^~~~~~~~~~~~~~~~~
> ../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
> 111 | __attach_tp(tp_xdp_redirect_map);
> | ^~~~~~~~~~~
> xdp_redirect_devmap_multi.c:37:1: note: in expansion of macro
‘DEFINE_SAMPLE_INIT’
> 37 | DEFINE_SAMPLE_INIT(xdp_redirect_devmap_multi);
> | ^~~~~~~~~~~~~~~~~~
> xdp_redirect_devmap_multi.c: In function ‘do_redirect_devmap_multi’:
> xdp_redirect_devmap_multi.c:108:24: error: implicit declaration of
function ‘bpf_object__next_program’; did you mean
‘bpf_object__unpin_programs’? [-Werror=implicit-function-declaration]
> 108 | while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
> | ^~~~~~~~~~~~~~~~~~~~~~~~
> | bpf_object__unpin_programs
> xdp_redirect_devmap_multi.c:108:22: error: assignment to ‘struct
bpf_program *’ from ‘int’ makes pointer from integer without a cast
[-Werror=int-conversion]
> 108 | while ((prog = bpf_object__next_program(skel->obj, prog)) != NULL)
> | ^
> xdp_redirect_devmap_multi.c:110:21: error: implicit declaration of
function ‘bpf_program__expected_attach_type’; did you mean
> bpf_program__set_expected_attach_type’?
[-Werror=implicit-function-declaration]
> 110 | bpf_program__expected_attach_type(prog) == BPF_XDP)
> | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> | bpf_program__set_expected_attach_type
> cc1: all warnings being treated as errors
> In file included from xdp_basic.c:25:
> xdp_basic.c: In function ‘sample_init’:
> ../lib/util/xdp_sample.h:72:21: error: implicit declaration of function
‘bpf_program__type’; did you mean ‘bpf_program__size’?
[-Werror=implicit-function-declaration]
> 72 | if (bpf_program__type(skel->progs.name) != BPF_PROG_TYPE_TRACING)
> | ^~~~~~~~~~~~~~~~~
> ../lib/util/xdp_sample.h:111:25: note: in expansion of macro ‘__attach_tp’
> 111 | __attach_tp(tp_xdp_redirect_map);
> | ^~~~~~~~~~~
> xdp_basic.c:30:1: note: in expansion of macro ‘DEFINE_SAMPLE_INIT’
> 30 | DEFINE_SAMPLE_INIT(xdp_basic);
> | ^~~~~~~~~~~~~~~~~~
> cc1: all warnings being treated as errors
> make[1]: *** [../lib/common.mk:107: xdp-bench] Error 1
> make: *** [Makefile:34: xdp-bench] Error 2
> """
>
> how to resolve this issue
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message
ID: ***@***.***>

--- a/configure
+++ b/configure
@@ -278,6 +278,7 @@ check_libbpf_functions()
     check_libbpf_function "bpf_program__insn_cnt" "(NULL)" "" "$LIBBPF_CFLAGS" "$LIBBPF_LDLIBS"
     check_libbpf_function "bpf_program__type" "(NULL)" "" "$LIBBPF_CFLAGS" "$LIBBPF_LDLIBS"
     check_libbpf_function "bpf_program__flags" "(NULL)" "" "$LIBBPF_CFLAGS" "$LIBBPF_LDLIBS"
+    check_libbpf_function "bpf_program__expected_attach_type" "(NULL)" "" "$LIBBPF_CFLAGS" "$LIBBPF_LDLIBS"
     check_libbpf_function "bpf_map_create" "(0, NULL, 0, 0, 0, NULL)" "" "$LIBBPF_CFLAGS" "$LIBBPF_LDLIBS"
     check_libbpf_function "perf_buffer__new_raw" "(0, 0, NULL, NULL, NULL, NULL)" "" "$LIBBPF_CFLAGS" "$LIBBPF_LDLIBS"
     check_libbpf_function "bpf_xdp_attach" "(0, 0, 0, NULL)" "" "$LIBBPF_CFLAGS" "$LIBBPF_LDLIBS"

--- a/lib/testing/test-tool.c
+++ b/lib/testing/test-tool.c
@@ -15,19 +15,10 @@
 #include "logging.h"
 #include "util.h"
 #include "xdp_sample.h"
+#include "compat.h"
 
 
 #define PROG_NAME "test-tool"
-
-
-#ifndef HAVE_LIBBPF_BPF_OBJECT__NEXT_PROGRAM
-static struct bpf_program *bpf_object__next_program(const struct bpf_object *obj,
-  						    struct bpf_program *prog)
-{
-  	return bpf_program__next(prog, obj);
-}
-#endif
-
 
 struct enum_val xdp_modes[] = {
        {"native", XDP_MODE_NATIVE},

--- a/lib/util/compat.h
+++ b/lib/util/compat.h
@@ -1,8 +1,11 @@
 #ifndef __COMPAT_H
 #define __COMPAT_H
 
+#include <bpf/btf.h>
+#include <bpf/libbpf.h>
+
 #ifndef HAVE_LIBBPF_BTF__TYPE_CNT
-static __u32 btf__type_cnt(const struct btf *btf)
+static inline __u32 btf__type_cnt(const struct btf *btf)
 {
 	/* old function didn't include 'void' type in count */
 	return btf__get_nr_types(btf) + 1;

--- a/lib/util/compat.h
+++ b/lib/util/compat.h
@@ -9,4 +9,26 @@ static __u32 btf__type_cnt(const struct btf *btf)
 }
 #endif
 
+#ifndef HAVE_LIBBPF_BPF_PROGRAM__TYPE
+static inline enum bpf_prog_type bpf_program__type(const struct bpf_program *prog)
+{
+	return bpf_program__get_type((struct bpf_program *)prog);
+}
+#endif
+
+#ifndef HAVE_LIBBPF_BPF_OBJECT__NEXT_PROGRAM
+static inline struct bpf_program *bpf_object__next_program(const struct bpf_object *obj,
+							   struct bpf_program *prog)
+{
+	return bpf_program__next(prog, obj);
+}
+#endif
+
+#ifndef HAVE_LIBBPF_BPF_PROGRAM__EXPECTED_ATTACH_TYPE
+static inline enum bpf_attach_type bpf_program__expected_attach_type(const struct bpf_program *prog)
+{
+	return bpf_program__get_expected_attach_type(prog);
+}
+#endif
+
 #endif

--- a/lib/util/xdp_sample.h
+++ b/lib/util/xdp_sample.h
@@ -6,6 +6,7 @@
 #include <getopt.h>
 
 #include <xdp/xdp_sample_shared.h>
+#include "compat.h"
 
 enum stats_mask {
 	_SAMPLE_REDIRECT_MAP         = 1U << 0,

--- a/xdp-trafficgen/xdp_trafficgen.bpf.c
+++ b/xdp-trafficgen/xdp_trafficgen.bpf.c
@@ -16,6 +16,16 @@
 #include <xdp/xdp_sample_common.bpf.h>
 #include <xdp/parsing_helpers.h>
 
+#if defined(HAVE_LIBBPF_BPF_PROGRAM__FLAGS) && defined(DEBUG)
+/* We use the many-argument version of bpf_printk() for debugging, so only
+ * enable it if we have the libbpf helper that selects the vprintf version. This
+ * was introduced in libbpf 0.6.0, which is the same versionn as the
+ * bpf_program__flags() method, so use that as an indicator since we don't
+ * feature detect on the BPF helpers themselves.
+ */
+#define TCP_DEBUG
+#endif
+
 char _license[] SEC("license") = "GPL";
 
 struct {
@@ -202,7 +212,7 @@ int xdp_handle_tcp_recv(struct xdp_md *ctx)
 	}
 
 	ack_seq = bpf_ntohl(tcphdr->ack_seq);
-#ifdef DEBUG
+#ifdef TCP_DEBUG
 	bpf_printk("Got state seq %u ack_seq %u new %u seq %u new %u window %u\n",
 		   fstate->seq,
 		   fstate->ack_seq, ack_seq,
@@ -246,7 +256,7 @@ int xdp_redirect_send_tcp(struct xdp_md *ctx)
 	struct tcphdr *tcphdr;
 	struct datarec *rec;
 	__u8 resend = 0;
-#ifdef DEBUG
+#ifdef TCP_DEBUG
 	__u8 print = 0;
 #endif
 	__u16 pkt_len;
@@ -292,7 +302,7 @@ int xdp_redirect_send_tcp(struct xdp_md *ctx)
 	new_seq = fstate->seq;
 	ack_seq = fstate->ack_seq;
 	window = fstate->window << fstate->wscale;
-#ifdef DEBUG
+#ifdef TCP_DEBUG
 	if (fstate->last_print != fstate->seq) {
 		fstate->last_print = fstate->seq;
 		print = 1;
@@ -305,7 +315,7 @@ int xdp_redirect_send_tcp(struct xdp_md *ctx)
 		 */
 		action = XDP_DROP;
 		bpf_spin_unlock(&fstate->lock);
-#ifdef DEBUG
+#ifdef TCP_DEBUG
 		if (print)
 			bpf_printk("Dropping because %u isn't before %u (ack_seq %u wnd %u)",
 				   new_seq + pkt_len, ack_seq + window, ack_seq, window);

